src/marlin_stubs/sdcard/M20-M30_M32-M34.cpp

```diff
@@ -1,22 +1,77 @@
 #include <dirent.h>
 
 #include "../../lib/Marlin/Marlin/src/gcode/gcode.h"
+#include "libs/hex_print_routines.h"
 #include "marlin_server.hpp"
 #include "media.h"
 #include "marlin_vars.hpp"
 
-// M20 - List SD card
-void GcodeSuite::M20() {
-    SERIAL_ECHOLNPGM(MSG_BEGIN_FILE_LIST);
+static void print_file_size(const char *const filename) {
+    struct stat stat_buf;
+    int rc = stat(filename, &stat_buf);
+    if (rc == 0) {
+        SERIAL_ECHO(stat_buf.st_size);
+    }
+}
+
+static void print_file_info(const char *const dirname, const char *const long_dirname, bool timestamps, bool long_filenames) {
+    SERIAL_ECHO(&dirname[5]);
+    SERIAL_CHAR(' ');
+    print_file_size(dirname);
+    if (timestamps) {
+        struct stat stat_buf;
+        int rc = stat(dirname, &stat_buf);
+        if (rc == 0) {
+            time_t t = stat_buf.st_mtim.tv_sec;
+            struct tm lt;
+            localtime_r(&t, &lt);
+            uint16_t m20_date = (lt.tm_year + 1900 - 1980) << 9 | (lt.tm_mon + 1) << 5 | lt.tm_mday;
+            uint16_t m20_time = lt.tm_hour << 11 | lt.tm_min << 5;
+            m20_time |= int((lt.tm_sec - (lt.tm_sec % 2)) / 2);
```

Comment:
Isn't this just `(lt.tm_sec >> 1)`? Entire thing would deserve to be factored out and unit tested. Why is there `1900 - 1980` ?

And I am also not sure if localtime_r is right function to call, but I was not looking into that too much. Just noting that timezones are always fun to deal with.

Reply:
This is taken from octoprint and tested. 1900 is the basis of timestruct whereas the value expected is value -1980 which would come from the fat entry. Tested to be correct, so I'll leave it this way.

Reply:
Done.